tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,329 @@
+    J. Chem. Phys. 107, 6756 (1997).
+
+    Example
+    -------
+    t = np.linspace(0, 10000, 10000)
+    f1 = 2*np.pi*0.1 - 1j*0.002
+    f2 = 2*np.pi*0.2 - 1j*0.0005
+    sig = 2*np.exp(-1j*f1*t) + 3*1j*np.exp(-1j*f2*t)
+    resdata = ResonanceFinder(freq_window=(0.05, 0.25)).run(sig, 1)
+    """
+
+    freq_window: Tuple[float, float] = Field(
+        ...,
+        title="Window ``[fmin,fmax]``",
+        description="Window ``[fmin,fmax]`` for the initial frequencies. "
+        "The resonance finder is initialized with an even grid of frequencies between "
+        "fmin and fmax. The resonance finder then iteratively optimizes and prunes these "
+        "frequencies. Note that frequencies outside this window may be returned.",
+        units=HERTZ,
+    )
+
+    num_freqs_bounds: Tuple[PositiveInt, PositiveInt] = Field(
+        (20, 300),
+        title="Bounds ``[nmin, nmax]`` on the initial number of frequencies.",
+        description="Bounds ``[nmin, nmax]`` on the initial number of frequencies. "
+        "The resonance finder is initialized with ``max(nmin, min(nmax, nguess))`` frequencies, "
+        "where ``nguess = signal_length*time_step*(fmax - fmin)/2``. The resonance finder then "
+        "iteratively optimizes and prunes these frequencies. Note that the number of frequencies "
+        "returned may be less than ``nmin`` but will not exceed ``nmax``. "
+        "Making these bounds larger can find more resonances at the expense of run-time "
+        "and matrix conditioning. ",
+    )
+
+    rcond: NonNegativeFloat = Field(
+        1e-4,
+        title="Cutoff for eigenvalues",
+        description="Cutoff for eigenvalues, relative to the largest eigenvalue. "
+        "The algorithm requires solving a generalized eigenvalue problem of the form "
+        "Ax = lambda B x. If B has small eigenvalues, this is poorly conditioned, "
+        "so we eliminate eigenvalues of B less than ``rcond`` times the largest eigenvalue. "
+        "Making this closer to one will typically return fewer resonances, while making it "
+        "closer to zero will typically return more resonances.",
+    )
+
+    @validator("freq_window", always=True)
+    def _check_freq_window(cls, val):
+        """Validate freq_window"""
+        if val[1] < val[0]:
+            raise ValidationError("The value of fmax is less than the value of fmin.")
+        return val
+
+    @validator("num_freqs_bounds", always=True)
+    def _check_num_freqs_bounds(cls, val):
+        """Validate freq_window"""
+        if val[1] < val[0]:
+            raise ValidationError("The value of nmax is less than the value of nmin.")
+        return val
+
+    def run(self, signal: ArrayLike[complex, 1], time_step: float) -> xr.Dataset:
+        """Finds resonances in a time series.
+        Note that the signal should start after the sources have turned off.
+
+        Parameters
+        ----------
+        signal : ArrayLike[complex, 1]
```

Comment:
the user unfortunately doesn't know what the `ArrayLike` type refers to. So I think we should just leave this as type of `List[complex]` and then immediately convert it to numpy array, ie `signal = np.ndarray(signal)`. perhaps with a subsequent heck that `len(signal.shape) == 1` or raise error telling them that the signal should be 1D.

In the description, perhaps change to 
```
one-dimensional array holding the complex-valued time series data to search for resonances
```
to be explicit.